src/hotspot/share/oops/compressedOops.inline.hpp

```diff
@@ -48,6 +48,9 @@ inline oop CompressedOops::decode_raw_not_null(narrowOop v) {
 }
 
 inline oop CompressedOops::decode_raw(narrowOop v) {
+  // Assume a null base casts to zero.  Otherwise we need more complexity that
+  // we can't test, since this is true for all currently supported platforms.
+  assert(0 == reinterpret_cast<uintptr_t>(nullptr), "null pointer value not zero?");
```

Comment:
Could this be a static_assert?

Reply:
+1. Although I would expect any sane compiler to fold it, maybe it is still not optimized with something like `-O0`. Or maybe just move these asserts to `CompressedOops::initialize`, so whatever happens, happens once.

Reply:
Correct that static_assert can't be used, because of the reinterpret_cast.

I considered putting the assert in CompressedOops::set_base, with comments to
connect encode/decode with that assertion, but prefer putting the assert near
the code that is actually relying on the property. Even with -O0 gcc seems to
constant fold the expression and (pretty much?) compile away the assertion.

A null pointer value is not guaranteed to have a zero representation.  The
conversion of a literal 0 to a null pointer value is syntactic sugar.  The
bit pattern of the result might be something else.  There's some good
discussion of this here:
https://stackoverflow.com/questions/2761360/could-i-ever-want-to-access-the-address-zero

Reply:
All right, if this assert folds, this all is just nitpicking then. Ship it.